ovs-dpdk.md

```diff
@@ -0,0 +1,170 @@
+Available modules are: uio_pci_generic, igb_uio and vfio-pci.
+In many cases, the standard uio_pci_generic module included in the Linux kernel can provide the uio capability.
+For some devices which lack support for legacy interrupts, igb_uio should be used, but it’s an out-of-tree kernel module.
+VFIO is preferred to uio drivers in latest DPDK versions in platforms that support VFIO.
+
+All ports that you would like to run DPDK application should be bound to one of the modules before OVS is started. DPDK will ignore all devices which are not bound to one of the drivers above.
+ 
+Install drivectl:
+```
+# yum install driverctl
+```
+Load vfio-pci
+
+```
+# modprobe vfio_pci
+```
+Review the status of current port bindings:
+```
+# driverctl -v list-devices
+```
+Bind the device to DPDK:
+```
+# driverctl set-override <PCI_ID> vfio-pci
+```
+
+### Hugepages Support
+Hugepages support is required for increasing performance. By allocating hugepages, we gain larger pages, using less pages and therefore less TLB misses are involved which speeds up performance. For these reasons, 1GB hugepages will be favored.
+Number of pages can vary according to usage requirements. 
+Hugepages must be marked as shared in order to enable host and guest communication, as packets are transmitted and received on buffers allocated on hugepages memory. 
+
+Add engine custom properties for hugepages support (followed by engine restart)
+```
+# engine-config -s "UserDefinedVMProperties=hugepages=^.*$;hugepages_shared=(true|false)"
+```
+ 
+
+### Enable DPDK in OVS:
+
+ovs-vscwhitchd should get a few parameters in order to enable DPDK functionality.
+
+Set dpdk-init so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+
+* dpdk-socket-mem: a comma seprated list of amount of memory to allocate from each NUMA node.
+For example, in a dual socket system, we allocate 1GB from NUMA node 0 and 0GB from the second:
+
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0" 
+```
+
+* dpdk-lcore-mask: Specifies the CPU cores on which dpdk lcore threads should be spawned and expects hex string.
+
+* pmd-cpu-mask: PMD (poll mode driver) threads are used to poll DPDK devices for new packets instead of using using the NIC driver that interacts with the kernel. 
+To better scale the workloads across cores, multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. Ensure that cores from the same NUMA nodes are enabled.
+eg: To spawn 2 pmd threads and pin them to cores 1, 2:
+
+```
+# ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+* Add the following entry to  /etc/sysconfig/openvswitch:
+```
+"DPDK_OPTIONS='-c coremask 100010 -n 2 --socket-mem 1024,0”
+```
+The pmd-cpu-mask and -c coremask have to be both specified, with the same value.
```

Comment:
sorry I still don't get it. `pmd-cpu-mask` is set to 6 (110 in binary) and `coremask` is 100010